nodes/modifier_make/cut_object_by_surface.py

```diff
@@ -117,51 +125,82 @@ def remember_connection(edges_by_face, bm_edge, bm_intersection):
+                        distance_limit -= 1e-6
+                        cast = bvh.ray_cast(start, edge, distance_limit)
+                        intersection = cast[0]
+                        distance = cast[3]
 
             bm_cut.verts.index_update()
 
             # Make "cut surface" object.
             # Of each object's face we are cutting, make an edge.
             bm_cut_edges = []
             for intersections in edges_by_face.values():
-                if len(intersections) == 1:
-                    self.debug("Only one of object face's edges intersects the surface")
+                if len(intersections) != 2:
+                    # self.debug("More or less than two intersection between object face's edges and surface")
+                    
+                    # Idea on how to deal with any number of intersections :
+                    # Intersect (project with raycast) each edge of the object with the surface mesh (already done)
+                    # Intersect each edge of the surface with the object mesh
+                    # For each intersection point, memorize to which faces it belongs,
+                    # i. e. the faces linked by the edge, and the face intersected by the edge
+                    # Save that information in a structure allowing easy access by faces
+                    # For each pair of (face from the object, face from the surface),
+                    # consider all the intersection points that belongs to both of them
+                    # There should be an even number of them, forming a line.
+                    # Link them two by two with an edge, starting from one extremity (1 with 2, 3 with 4, etc...)
+                    # Now, all the new edges should form rings, which can then be used to cut the surface and object meshes,
+                    # as we know which of their original edges/faces are concerned by which new edges
+                    # Finally, for each object part, the cut surface within the corresponding ring
+                    # is merged by distance with the part
+                    
+                    # This works in the general cases, but there are still a few corner cases to deal with :
+                    # If the surface and the object do not fully intersect each other,
+                    # i. e. there are new edges not included in any rings
+                    # If the surface or the object has wire geometry, or has more than two faces linked to an edge,
+                    # or has vertices linked to multiple mesh regions (see Blender select non manifold function)
                     continue
-                if len(intersections) > 2:
-                    # Some day we may support splitting the face in more than two pieces;
-                    # but the algorithm will be more complicated then.
-                    raise Exception("Unsupported: more than two edges of the face intersect the surface")
```

Comment:
I'd suggest to add an option to the node, how to proceed in this case: raise an exception or just continue. Because, if objects are generated somehow, it can be quite unexpected if with some values of parameters this node generates cut pieces, and with others it does not — if the setup is designed so that the objects should conform to restriction of this node, then, if for some reason they do not, the error would indicate the problem somewhere in objects generation routines, instead of output mysteriously disappearing.

Reply:
Corrected in new commit.